saleor/graphql/product/mutations/channels.py

```diff
@@ -156,12 +176,69 @@ def add_channels(cls, product: "ProductModel", add_channels: List[Dict]):
                     defaults["available_for_purchase"] = datetime.date.today()
                 else:
                     defaults["available_for_purchase"] = available_for_purchase_date
-            ProductChannelListing.objects.update_or_create(
+            product_channel_listing, _ = ProductChannelListing.objects.update_or_create(
                 product=product, channel=channel, defaults=defaults
             )
+            cls.add_variants(channel, add_variants)
+            cls.remove_variants(
+                product_channel_listing, product, channel, remove_variants
+            )
 
     @classmethod
-    def remove_channels(cls, product: "ProductModel", remove_channels: List[int]):
+    def validate_variants(cls, input, errors):
+        for update_channel in input.get("update_channels", []):
+            error = check_for_duplicates(
+                update_channel, "add_variants", "remove_variants", "variants"
+            )
+            if error:
+                error.code = ProductErrorCode.DUPLICATED_INPUT_ITEM.value
+                errors["addVariants"].append(error)
+
+    @classmethod
+    def add_variants(cls, channel, add_variants: List[Dict]):
+        if not add_variants:
+            return
+        variants_id = cls.get_nodes_or_error(add_variants, "id", ProductVariant)
+        variant_channel_listings = []
+        for variant in variants_id:
+            variant_channel_listings.append(
+                ProductVariantChannelListing(channel=channel, variant=variant)
+            )
```

Comment:
For me it's misleading, I'm expecting `ids` and got instances. IMO it should be changed to:
```suggestion
        variants = cls.get_nodes_or_error(add_variants, "id", ProductVariant)
        variant_channel_listings = []
        for variant in variants:
            variant_channel_listings.append(
                ProductVariantChannelListing(channel=channel, variant=variant)
            )
```